### PR TITLE
Add HTTP(S) proxy support

### PR DESCRIPTION
**Pull request by allquixotic**
This seems to work with both HTTP and HTTPS sites when using the jsonrpsee_http_client crate, allowing you to put a proxy in between your client and the endpoint. Very useful for debugging. Works with Proxyman, the proxy I'm using.

### CODE REVIEW

File: client/http-client/src/client.rs
Comment (niklasad1) [inline client/http-client/src/client.rs:82]:
```suggestion
	proxy: Option<Uri>,
```

File: client/http-client/Cargo.toml
Comment (niklasad1) [inline client/http-client/Cargo.toml:17]: it should be possible to disable tls and not bring all these deps

```suggestion
hyper-proxy = { version = "0.9", default-features = false }
```

File: client/http-client/src/transport.rs
Comment (niklasad1) [inline client/http-client/src/transport.rs:97]:
```suggestion
						let proxy_uri = pr.parse().map_err(|e| Error::Url(format!("Invalid proxy URI: {}", e)))?;
```

File: client/http-client/src/client.rs
Comment (niklasad1) [inline client/http-client/src/client.rs:141]: would be better to return `Result<Self, Error>` and to parse proxy URI directly here.

```rust
	pub fn set_proxy(mut self, proxy: impl AsRef<str>) -> Result<Self, Error> {
      self.proxy = proxy.parse().map_err(|e| Error::Transport(....))?;
      self
   }
```

File: client/http-client/src/transport.rs
Comment (niklasad1) [inline client/http-client/src/transport.rs:95]: needless clone

```suggestion
				match proxy {
```

File: client/http-client/src/transport.rs
Comment (niklasad1) [inline client/http-client/src/transport.rs:97]: if the proxy URI is parsed in the builder this hasn't to be done again.

```suggestion
```

File: client/http-client/src/transport.rs
Comment (niklasad1) [inline client/http-client/src/transport.rs:99]: we probably need to add `IoError` for this to work but no unwraps are allowed here.

```suggestion
						let proxy_connector = ProxyConnector::from_proxy(connector, proxy_obj)?;
```

**Review by niklasad1 — CHANGES_REQUESTED** [review]
Thank you, looks generally good but I have some comments/suggestion that needs to addressed before merging.

Also, this needs to be added to the `WS client` which I reckon would be useful as well.

**Comment by allquixotic** [comment]
> Also, this needs to be added to the WS client which I reckon would be useful as well.

Agreed, but this is PR is for HTTP proxy for the HTTP client; at least, that's what I'm able/willing to work on (I'm not using the lib with a WebSocket server). Will you consider merging this just for http if the code is corrected?

**Comment by allquixotic** [comment]
Weird; Clippy seems to be crashing in the compiler?

**Comment by niklasad1** [comment]
> Agreed, but this is PR is for HTTP proxy for the HTTP client; at least, that's what I'm able/willing to work on (I'm not using the lib with a WebSocket server). Will you consider merging this just for http if the code is corrected?

yeah, we can take of the `ws client` in another PR.

> Weird; Clippy seems to be crashing in the compiler?

yeah, it's known just ignore it for now https://github.com/rust-lang/rust-clippy/issues/9922

**Review by niklasad1 — APPROVED** [review]

File: client/http-client/src/client.rs
Comment (jsdw) [inline client/http-client/src/client.rs:88]: It looks like this is returned in an `Error`, so I think it should be `pub` so that it can be named by people.

Comment (jsdw) [reply client/http-client/src/client.rs:88]: (and probably it should also be moved to live alongside `Error`? It feels a bit out of place here)

Comment (jsdw) [reply client/http-client/src/client.rs:88]: Ah I see now' it's converted into an `anyhow::Error` before being returned; ignore me then :)

**Review by jsdw — APPROVED** [review]

**Review by lexnv — APPROVED** [review]

File: client/http-client/src/transport.rs
Comment (niklasad1) [inline client/http-client/src/transport.rs:40]:
```suggestion
			#[cfg(all(feature = "proxy", feature = "tls"))]
```

File: client/http-client/src/client.rs
Comment (niklasad1) [inline client/http-client/src/client.rs:152]:
```suggestion
	#[cfg(all(feature = "proxy", feature = "tls"))]
```

**Review by niklasad1 — CHANGES_REQUESTED** [review]
Hey again,

I have looked into this crate `hyper-proxy` and it doesn't really seems to be actively maintained and brings in a bunch of outdated dependencies such as `hyper-rustls`, https://github.com/tafia/hyper-proxy/pull/28/files has been opened over a year.

I think we should find another library and or write something ourselves using tower middleware.
Sorry, figured this out before I was going to merge it.

**Comment by allquixotic** [comment]
Created [issue hyper#3124](https://github.com/hyperium/hyper/issues/3124) to ask the hyper maintainers what they would suggest for HTTP proxy support. I'd rather not reinvent it within jsonrpsee. 

Supporting an http proxy within an HTTP client library is a generally useful feature that can and should be either part of the hyper crate itself, or part of a maintained crate focused on that feature. This is basic stuff that exists everywhere else I've done HTTP requests (JavaScript, .NET, Java, GLib/C, Qt/C++, Python, Ruby, etc.) It's not jsonrpsee's responsibility to code this up.

**Pull request closed without merging**
